crud/src/main/java/com/redhat/lightblue/crud/mongo/Translator.java

```diff
@@ -380,21 +381,31 @@ public DBObject translate(EntityMetadata md, UpdateExpression expr)
                                               Sort s) {
         Set<Path> fields=new HashSet<>();
         FieldCursor cursor=md.getFieldCursor();
-        while(cursor.next()) {
-            Path field=cursor.getCurrentPath();
-            FieldTreeNode node=cursor.getCurrentNode();
-            if( (node instanceof ObjectField) ||
-                (node instanceof ArrayField &&
-                 ((ArrayField)node).getElement() instanceof ObjectArrayElement) ) {
-                // include its member fields
-            } else if( (p!=null && p.isFieldRequiredToEvaluateProjection(field)) ||
-                       (q!=null && q.isRequired(field)) ||
-                       (s!=null && s.isRequired(field)) ) {
-                LOGGER.debug("{}: required",field);
-                fields.add(field);
-            } else {
-                LOGGER.debug("{}: not required", field);
-            }
+        if(cursor.next()) {
+            boolean done=false;
+            do {
+                Path field=cursor.getCurrentPath();
+                FieldTreeNode node=cursor.getCurrentNode();
+                LOGGER.debug("Checking if {} is included ({})",field,node);
+                if(node instanceof ResolvedReferenceField||
+                   node instanceof ReferenceField) {
+                    done=!cursor.nextSibling();
```

Comment:
It shouldn't stop early. cursor.next moves depth first. When we see a
resolvedReference, it calls nextSibling, i.e. should skip the whole subtree
rooted at that reference, and continue depth first.

On Wed, Sep 16, 2015 at 9:24 AM, Naveen Malik <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/crud/mongo/Translator.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/153#discussion_r39644022>
> :
>
> > -                       (q!=null && q.isRequired(field)) ||
> > -                       (s!=null && s.isRequired(field)) ) {
> > -                LOGGER.debug("{}: required",field);
> > -                fields.add(field);
> > -            } else {
> > -                LOGGER.debug("{}: not required", field);
> > -            }
> > +        if(cursor.next()) {
> > +            boolean done=false;
> > +            do {
> > +                Path field=cursor.getCurrentPath();
> > +                FieldTreeNode node=cursor.getCurrentNode();
> > +                LOGGER.debug("Checking if {} is included ({})",field,node);
> > +                if(node instanceof ResolvedReferenceField||
> > +                   node instanceof ReferenceField) {
> > +                    done=!cursor.nextSibling();
>
> Couldn't this stop too early? What if the ref field is in the middle of
> the cursor? What if there are multiple references in the query?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/153/files#r39644022>
> .
>